src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,436 @@
+    /// @param from The address of the one creating the game (may be different from msg.sender if metaTx)
+    /// @param gameId The id of the GAME to add asset to
+    /// @param assetIds The id of the asset to add to GAME
+    /// @param values The amount of each asset to add to GAME
+    function addMultipleAssets(
+        address from,
+        uint256 gameId,
+        uint256[] memory assetIds,
+        uint256[] memory values,
+        string memory uri
+    ) public override minterGuard() onlyOwnerOrEditor(gameId) {
+        require(assetIds.length == values.length, "INVALID_INPUT_LENGTHS");
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            _gameData[gameId]._assets.add(assetIds[i]);
+            uint256 assetValues = _gameData[gameId]._values[assetIds[i]];
+            _gameData[gameId]._values[assetIds[i]] = assetValues.add(values[i]);
+        }
+        _asset.safeBatchTransferFrom(from, address(this), assetIds, values, "");
+
+        setTokenURI(gameId, uri);
+        emit AssetsAdded(gameId, assetIds, values);
+    }
+
+    /// @notice Set the URI of a specific game token
+    /// @param gameId The id of the game token
+    /// @param URI The URI string for the token's metadata
+    function setTokenURI(uint256 gameId, string memory URI) public override {
+        require(
+            // @review !
+            msg.sender == _ownerOf(gameId) || _gameEditors[gameId][msg.sender] || msg.sender == _minter,
+            "URI_ACCESS_DENIED"
+        );
+        _metaData[gameId] = URI;
+    }
+
+    /// @notice Function to get the numnber of assets in a game
+    /// @param gameId The is of the GAME to query
+    /// @return assetTypes The ids of each asset in the game
+    /// @return totalAssets The quantity of each asset ID in the game
+    function getNumberOfAssets(uint256 gameId) public view override returns (uint256 assetTypes, uint256 totalAssets) {
+        uint256 assets = _gameData[gameId]._assets.length();
+        uint256 total;
+        for (uint256 i = 0; i < assets; i++) {
+            total += _gameData[gameId]._values[_gameData[gameId]._assets.at(i)];
+        }
+        // get each value and add together
+        return (assets, total);
+    }
+
+    /// @notice Return the URI of a specific token
+    /// @param gameId The id of the token
+    /// @return uri The URI of the token
+    function tokenURI(uint256 gameId) public view override returns (string memory uri) {
+        require(_ownerOf(gameId) != address(0), "BURNED_OR_NEVER_MINTED");
+        string memory URI = _metaData[gameId];
+        return URI;
+    }
+
+    /// @dev Function to create a new gameId and associate it with an owner
+    /// @param to The address of the Game owner
+    /// @return id The newly created gameId
+    function _mintGame(address from, address to) internal returns (uint256 id) {
+        uint256 gameId = generateGameId(from);
+        _nextId = _nextId + 1;
```

Comment:
_nextId make it asier to mint new game

the disadvantage are 
- extra storage
- no way to predict the tokenId client side

For Asset we choose to go with the predictable tokenID and ask the user to give the tokenID
This means at minting time, the user need help to pick an unused tokenID

Reply:
Removed nextId and added a `randomId` param (uint96). This will be calculated on the back end and passed to the winter contract.

Reply:
Switched to` randomId` passed in to the createGame function.